core/lib/types/src/fee_model.rs

```diff
@@ -236,9 +239,85 @@ pub struct FeeParamsV1 {
 
 #[derive(Debug, Clone, Copy, Serialize, Deserialize)]
 pub struct FeeParamsV2 {
-    pub config: FeeModelConfigV2,
-    pub l1_gas_price: u64,
-    pub l1_pubdata_price: u64,
+    config: FeeModelConfigV2,
+    l1_gas_price: u64,
+    l1_pubdata_price: u64,
+    conversion_ratio: BaseTokenConversionRatio,
+}
+
+impl FeeParamsV2 {
+    pub fn new(
+        config: FeeModelConfigV2,
+        l1_gas_price: u64,
+        l1_pubdata_price: u64,
+        conversion_ratio: BaseTokenConversionRatio,
+    ) -> Self {
+        Self {
+            config,
+            l1_gas_price,
+            l1_pubdata_price,
+            conversion_ratio,
+        }
+    }
+
+    /// Returns the fee model config with the minimal L2 gas price denominated in the chain's base token (WEI or equivalent).
+    pub fn config(&self) -> FeeModelConfigV2 {
+        FeeModelConfigV2 {
+            minimal_l2_gas_price: self.convert_to_base_token(self.config.minimal_l2_gas_price),
+            ..self.config
+        }
+    }
+
+    /// Returns the l1 gas price denominated in the chain's base token (WEI or equivalent).
+    pub fn l1_gas_price(&self) -> u64 {
+        self.convert_to_base_token(self.l1_gas_price)
+    }
+
+    /// Returns the l1 pubdata price denominated in the chain's base token (WEI or equivalent).
+    pub fn l1_pubdata_price(&self) -> u64 {
+        self.convert_to_base_token(self.l1_pubdata_price)
+    }
+
+    /// Converts the fee param to the base token.
+    fn convert_to_base_token(&self, price_in_wei: u64) -> u64 {
+        let conversion_ratio = BigDecimal::from(self.conversion_ratio.numerator.get())
+            / BigDecimal::from(self.conversion_ratio.denominator.get());
+        let converted_price_bd = BigDecimal::from(price_in_wei) * conversion_ratio;
+
+        // Match on the converted price to ensure it can be represented as a u64
+        match converted_price_bd.to_u64() {
+            Some(converted_price) => converted_price,
+            None => {
+                if converted_price_bd > BigDecimal::from(u64::MAX) {
+                    tracing::warn!(
+                        "Conversion to base token price failed: converted price is too large: {}. Using u64::MAX instead.",
+                        converted_price_bd
+                    );
+                } else {
+                    tracing::error!(
+                    "Conversion to base token price failed: converted price is not a valid u64: {}. Using u64::MAX instead.",
+                    converted_price_bd
```

Comment:
This shouldn't be possible, right? The numerator, denominator, and gas price are all positive, and we've checked that the value is less than `u64::MAX`. So reaching this point would mean a big problem in the implementation. I'd say it's adequate to panic here, because we don't know what's going on and it's not safe to continue.